### DISCUSSION TITLE
PAGSurface readPixels失败

### DISCUSSION THREAD

**Discussion by krislyy** [Tencent/libpag#77] [Q&A] [answered]
```
private void pagToGif() {
        // pag player
        PAGComposition composition = makeComposition();
        int height = composition.height();
        int width = composition.width();
        PAGSurface surface = PAGSurface.MakeOffScreen(width, height);
        if (surface == null) return;
        PAGPlayer player = new PAGPlayer();
        player.setComposition(composition);
        player.setSurface(surface);
        player.setProgress(0);
        // byte buffer
        ByteBuffer buffer = ByteBuffer.allocateDirect(width * height * 4)
                .order(ByteOrder.nativeOrder());
        buffer.clear();
        // 开始导出
        int totalFrames = (int) (composition.duration() * composition.frameRate() / 1000000);
        for (int i = 0; i < totalFrames; i++) {
            generateSurfaceFrame(player, composition, i);
            surface.readPixels(buffer);
            Bitmap bitmap = BitmapFactory.decodeByteArray(buffer.array(), 0, buffer.capacity());
            buffer.clear();
        }
    }
```
PAGSurface的readPixels函数返回的buffer一直空的，想问下Android需要特殊设置什么才能拿到当前渲染的到surface上图片呢？

**Comment by domchen** [comment, marked as answer]
你这边关键的步骤代码几乎都没贴出来，PAGComposition是怎么创建的？是不是一个空的内容。首帧是空的内容吗？generateSurfaceFrame里做什么了，有调用flush吗？ PAGSurface.MakeOffscreen()是哪来的？android不存在这个方法。surface.readPixels(buffer); 这个又是哪来的方法？最后你是要readPixels，为啥read完了你把它当做文件去解码。你这个的代码是编译通过的吗？

**Reply by krislyy** [reply]
```
Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
bitmap.copyPixelsFromBuffer(buffer);
```
写的有点问题，应该是创建bitmap。然后就能读出数据了。离屏相关的是我自己暴露出了源码中的接口重新编译的。 现在能拿到图片了，另外我想问下，如果我只是单纯的想导出gif，有没有更好的方案提供呢？

**Reply by domchen** [reply]
导出gif我们也是用离屏readPixels的方式做的。PAGViewer里有个导出APNG的能力，原理类似。不过我们都是在C++层面操作会简单很多。你这个感觉也没有必要一个个去暴露接口，直接用C++接口写个组件，统一暴露一个Java接口传出编码完的gif结果就好。

**Reply by krislyy** [reply]
我用readPixels读取一帧耗时500多毫秒，再加上编码成png耗时300多，整个编译一帧耗时太久了，不知道你们导出一帧 APNG耗时怎么样？ps:我用的手机华为p20。 另外，我现在采用ImageReader方式读取的话平均一帧耗时10ms，我想接下来就是找一个比较快的gif编码算法就好了。

**Reply by domchen** [reply]
我印象中应该没有那么高，你可以检查一下是否存在多次内存拷贝的操作。直接按最终需要的格式从C++层读取出来，在C++层直接编码，这样避免多次数据在C++和Java之间来回拷贝应该会节省一些时间。
